kubernetes/cache/kube_cache.go

```diff
@@ -298,8 +331,17 @@ func (c *kubeCache) refresh(namespace string) error {
 		namespace = ""
 	}
 
-	c.stop(namespace)
-	return c.startInformers(namespace)
+	// We need to be thread-safe here so refreshing doesn't deadlock after a deleted namespace is re-created.
+	// Locking here ensures concurrent calls to getCacheLister doesn't deadlock. There tends to be a flurry
+	// of concurrent calls to getCacheLister and if at that time the namespace cache needs to be refreshed, we
+	// only want to create informers once here. This might not always create informers one time, but it helps limit
+	// the number of times informers are created and then immediately thrown away and created again.
```

Comment:
So we could end up starting multiple informers for the same namespace/resource?

Reply:
Can we put a `refreshLock` on `getCacheLister` that does a 
```
c.refreshLock.RLock()
defer c.refreshLock.RUnlock()
```
and inside `refresh` we do:
```
c.refreshLock.Lock()
defer c.refreshLock.Unlock()
```
I something like that should help ensure that we don't try to read from the listers map while they are refreshing since callers won't be able to get a readlock while a refresh is happening.

Reply:
> So we could end up starting multiple informers for the same namespace/resource?

That's me doing a CYA :) I have not see multiple informers get created. The do-once stuff appears to be working as intended. But, because I was not 100% confident in my understanding of the thread-safety stuff surrounding this code, I made that statement.

I actually DID do what you suggested with the refresh lock - and was getting deadlocks consistently (hence why my confidence isn't 100% - I couldn't figure out why that deadlock was happening - I think there are some simultenous go func executions happening that I could not find). That's why there is this OnceWrapper - I went to the design to avoid the deadlocks BUT still ensure we only run it one time.

Reply:
I'm going to revisit the refresh lock thing, just to see if I see that deadlock again. This is the last thing left for me to look at - all other suggestions have been implemented in the latest commit.

Reply:
OK, I see the deadlocks like I saw before. Currently using this:
```diff
diff --git a/kubernetes/cache/kube_cache.go b/kubernetes/cache/kube_cache.go
index 5bf59ef93..326aaf5cc 100644
--- a/kubernetes/cache/kube_cache.go
+++ b/kubernetes/cache/kube_cache.go
@@ -177,7 +177,8 @@ type ErrorHandler func(kc *kubeCache, namespace string, err error)
 
 // kubeCache is a local cache of kube objects. Manages informers and listers.
 type kubeCache struct {
-       cacheLock sync.RWMutex
+       cacheLock   sync.RWMutex
+       refreshLock sync.RWMutex
        // refreshOnce enforces thread safety around the re-starting of informers
        refreshOnce        *OnceWrapper
        cfg                config.Config
@@ -318,6 +319,10 @@ func (c *kubeCache) stop(namespace string) {
 // Refresh will recreate the necessary cache. If the cache is cluster-scoped the "namespace" argument
 // is ignored and the whole cache is recreated, otherwise only the namespace-specific cache is updated.
 func (c *kubeCache) Refresh(namespace string) {
+       // COMMENT OUT BELOW AND IT STILL DEADLOCKS
+       c.refreshLock.Lock()
+       defer c.refreshLock.Unlock()
+
        c.cacheLock.Lock()
        defer c.cacheLock.Unlock()
 
@@ -327,6 +332,9 @@ func (c *kubeCache) Refresh(namespace string) {
 }
 
 func (c *kubeCache) refresh(namespace string) error {
+       c.refreshLock.Lock()
+       defer c.refreshLock.Unlock()
+
        if c.clusterScoped {
                namespace = ""
        }
@@ -612,6 +620,9 @@ func (c *kubeCache) getCacheLister(namespace string) *cacheLister {
                return c.clusterCacheLister
        }
 
+       c.refreshLock.RLock()
+       defer c.refreshLock.RUnlock()
+
        lister, ok := c.nsCacheLister[namespace]
        if !ok {
 ```

Reply:
This is the unit test that deadlocks after making that change: https://github.com/kiali/kiali/blob/v1.89.1/kubernetes/cache/kube_cache_test.go#L112

Reply:
Tried locking directly in the do-once function, still deadlocks:
```diff
diff --git a/kubernetes/cache/kube_cache.go b/kubernetes/cache/kube_cache.go
index 5bf59ef93..8d2bbcfc1 100644
--- a/kubernetes/cache/kube_cache.go
+++ b/kubernetes/cache/kube_cache.go
@@ -177,7 +177,8 @@ type ErrorHandler func(kc *kubeCache, namespace string, err error)
 
 // kubeCache is a local cache of kube objects. Manages informers and listers.
 type kubeCache struct {
-       cacheLock sync.RWMutex
+       cacheLock   sync.RWMutex
+       refreshLock sync.RWMutex
        // refreshOnce enforces thread safety around the re-starting of informers
        refreshOnce        *OnceWrapper
        cfg                config.Config
@@ -338,6 +339,8 @@ func (c *kubeCache) refresh(namespace string) error {
        // the number of times informers are created and then immediately thrown away and created again.
        defer c.refreshOnce.Reset()
        return c.refreshOnce.Do(func() error {
+               c.refreshLock.Lock()
+               defer c.refreshLock.Unlock()
                log.Debugf("Restarting cache informers for namespace [%v]", namespace)
                c.stop(namespace)
                return c.startInformers(namespace)
@@ -612,6 +615,9 @@ func (c *kubeCache) getCacheLister(namespace string) *cacheLister {
                return c.clusterCacheLister
        }
 
+       c.refreshLock.RLock()
+       defer c.refreshLock.RUnlock()
+
```

Reply:
I think it's happening here: https://github.com/kiali/kiali/blob/v1.89.1/kubernetes/cache/kube_cache.go#L337

Reply:
OK, if I release the lock after calling c.stop but before c.startInformers inside the "once.do" function, the deadlock goes away, but I don't think that is thread safe at that point.

```diff
        return c.refreshOnce.Do(func() error {
+               c.refreshLock.Lock()
                log.Debugf("Restarting cache informers for namespace [%v]", namespace)
                c.stop(namespace)
+               c.refreshLock.Unlock()
                return c.startInformers(namespace)
```

I see something odd about that startInformers function anyway. I need to stare at this some more.

Reply:
> I see something odd about that startInformers function anyway. I need to stare at this some more.

I will briefly explain here...

1. [startInformers](https://github.com/kiali/kiali/blob/v1.89.1/kubernetes/cache/kube_cache.go#L312-L338) runs go funcs to create the three sets of informers (kubernetes, istio, gateways)
2. [create istio informers](https://github.com/kiali/kiali/blob/v1.89.1/kubernetes/cache/kube_cache.go#L350-L357) and [create gateway informers](https://github.com/kiali/kiali/blob/v1.89.1/kubernetes/cache/kube_cache.go#L403-L410) both call [getCacheLister](https://github.com/kiali/kiali/blob/v1.89.1/kubernetes/cache/kube_cache.go#L535-L540) to get the listers they will work on
3. BUT the [createKubernetesInformers](https://github.com/kiali/kiali/blob/v1.89.1/kubernetes/cache/kube_cache.go#L441-L530) is the one that actually creates the listers (the ones that getCacheListers returns).

So how can the code expect (1) to work (i.e. simultaneously run the three create informers functions)  when one of them creates the listers the other 2 retrieve (??). That doesn't seem right. This might be working just by luck... the createKubernetesInformers needs to be guaranteed to run first; otherwise, the createIstio and createGateway functions calling getCacheLister aren't guaranteed to get listers back -- they are in a race with createKubernetes who is creating them.

Reply:
With the OnceWrapper in place, I don't think we have to worry about this refreshLock. We get thread-safety with OnceWrapper - I've seen it working properly and I've been checking this on purpose.